src/gridfs/upload.rs

```diff
@@ -173,3 +179,287 @@ impl GridFsBucket {
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsUploadStream {
+    /// Gets the file `id` for the stream.
+    pub fn id(&self) -> &Bson {
+        &self.id
+    }
+
+    pub async fn abort(&mut self) -> Result<()> {
+        match self.state {
+            State::Closed => Err(ErrorKind::GridFs(GridFsErrorKind::UploadStreamClosed).into()),
+            _ => {
+                self.bucket
+                    .chunks()
+                    .delete_many(doc! { "files_id": &self.id }, None)
+                    .await?;
+                self.state = State::Closed;
+                Ok(())
+            }
+        }
+    }
+}
+
+impl Drop for GridFsUploadStream {
+    // TODO RUST-1493: pre-create this task
+    fn drop(&mut self) {
+        if !matches!(self.state, State::Closed) {
+            let chunks = self.bucket.chunks().clone();
+            let id = self.id.clone();
+            runtime::execute(async move {
+                let _result = chunks.delete_many(doc! { "files_id": id }, None).await;
+            })
+        }
+    }
+}
+
+impl AsyncWrite for GridFsUploadStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::result::Result<usize, futures_io::Error>> {
+        let stream = self.get_mut();
+
+        let future = match &mut stream.state {
+            State::Idle(buffer) => {
+                let buffer_ref = buffer.as_mut().unwrap();
+
+                buffer_ref.extend_from_slice(buf);
+                if buffer_ref.len() < stream.chunk_size as usize {
+                    return Poll::Ready(Ok(buf.len()));
+                }
+
+                let new_future = write_bytes(
+                    stream.bucket.clone(),
+                    buffer.take().unwrap(),
+                    stream.current_n,
+                    stream.chunk_size,
+                    stream.id.clone(),
+                )
+                .boxed();
+                stream.state.set_writing(new_future)
+            }
+            State::Writing(future) => future,
+            State::Closing(_) | State::Closed => return Poll::Ready(Err(get_closed_error())),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        match result {
+            Ok((chunks_written, buffer)) => {
+                stream.current_n += chunks_written;
+                stream.state = State::Idle(Some(buffer));
+                Poll::Ready(Ok(buf.len()))
+            }
+            Err(error) => {
+                stream.state = State::Closed;
+                Poll::Ready(Err(error.into_futures_io_error()))
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        // The buffer only contains leftover bytes that couldn't fill an entire chunk, so there's
```

Comment:
Wouldn't writing such leftover bytes to the db be the correct usage of `flush`?

Reply:
Ah yeah I was imagining just having an irregularly sized chunk, but I remember now that we can't do this. Agreed that doing partial chunk writes doesn't seem worth it, so this seems fine as-is. In the documentation PR, it might be worth calling out that `flush` doesn't do anything due to this requirement.